### PR TITLE
Fix refresh token issue

### PR DESCRIPTION
**Pull request by manojmalik20**
#### Summary
Chimera didn't contain the functionality to refresh the OAuth token. This PR adds that functionality.

#### Related Links
https://github.com/mattermost/mattermost-plugin-gitlab/pull/298
https://github.com/mattermost/mattermost-plugin-gitlab/pull/298#issuecomment-1155996250

### CODE REVIEW

**Comment by DHaussermann** [comment]
Hi @Szymongib 
1. Are you able to provide review on this? I'm not sure if there is someone else who may own this. 
2. Do you have any idea on how this may be testable and who could help? Perhaps a change could be deployed only to the `test.mattermost.cloud` servers first? (I have no visibility into how this gets deployed and who does it)

**Comment by Szymongib** [comment]
Hey @DHaussermann,
Yes, I will have a look.

File: internal/api/handler_test.go [hunk 1]
```diff
@@ -504,6 +504,17 @@ func Test_HandleExchangeToken(t *testing.T) {
 		_ = assertRespStatus(t, http.DefaultClient, req, http.StatusBadRequest)
 	})
 
+	t.Run("return 400 if grant type is invalid", func(t *testing.T) {
+		form := url.Values{
+			"grant_type": {"abcd"},
+		}
+		req, err := http.NewRequest(http.MethodPost, proxyTokenURL, strings.NewReader(form.Encode()))
+		require.NoError(t, err)
```
Comment (Szymongib) [inline internal/api/handler_test.go:512]: This does not test the described case as `grant_type` is empty on the server-side. I think you need to add appropriate `Content-Type` for this to work.

File: internal/api/handler.go [hunk 2]
```diff
@@ -282,21 +282,60 @@ func (h *Handler) handleTokenExchange(c *OAuthAppContext, w http.ResponseWriter,
...
+			return
+		}
+
+		token, err = conf.Exchange(context.Background(), code)
+		if err != nil {
+			c.Logger.WithError(err).Error("Failed to exchange token")
+			http.Error(w, "failed to exchange token", http.StatusBadGateway)
+			return
+		}
+	case "refresh_token":
```
Comment (Szymongib) [inline internal/api/handler.go:316]: We should add a test for this case too.

**Review by Szymongib — COMMENTED** [review]
Thanks for the PR, looks good. I think we should add a unit test for the new case.

**Review by Szymongib — APPROVED** [review]
Thanks, @manojmalik20! Looks good!

**Comment by Szymongib** [comment]
@DHaussermann I am going to merge it and deploy it to the Test environment, after that you should be able to test with `test.mattermost.cloud` servers.

**Comment by DHaussermann** [comment]
@manojmalik20 I have tested this and I don't see the issue resolved.

I spun up a server with `/cloud` and still see the issue. I opened a new issue here. https://github.com/mattermost/chimera/issues/10

cc @Szymongib  unless I've misunderstood something - the change is being deployed when I spin up a cloud server now. Please advise.

**Comment by DHaussermann** [comment]
@Szymongib I did not test with that specific PR branch but I tested with the `v1.4.1` release https://github.com/mattermost/mattermost-plugin-gitlab/releases/tag/v1.4.1 which contains the same fix.

**Comment by Szymongib** [comment]
@DHaussermann could you share plugin logs? Does it error in any way?

**Comment by DHaussermann** [comment]
Hi @Szymongib 
yes, sorry I forgot to post the logs. 
```
{"timestamp":"2022-07-08 17:12:47.749 Z","level":"debug","msg":"Using Chimera Proxy OAuth configuration","caller":"app/plugin_api.go:934","plugin_id":"com.github.manland.mattermost-plugin-gitlab"}
{"timestamp":"2022-07-08 17:12:47.758 Z","level":"debug","msg":"Using Chimera Proxy OAuth configuration","caller":"app/plugin_api.go:934","plugin_id":"com.github.manland.mattermost-plugin-gitlab"}
{"timestamp":"2022-07-08 17:12:47.826 Z","level":"debug","msg":"Received HTTP request","caller":"web/handlers.go:156","method":"POST","url":"/api/v4/channels/members/me/view","request_id":"sbzqsxso97n6dqq7yzc5qafqxw","status_code":"200"}
{"timestamp":"2022-07-08 17:12:47.920 Z","level":"debug","msg":"Received HTTP request","caller":"web/handlers.go:156","method":"GET","url":"/api/v4/channels/88hq7ex5iff338yzh8nmjgey7c/posts","request_id":"a36sparu93d3jxgi7fqkf8b3ya","status_code":"200"}
{"timestamp":"2022-07-08 17:12:47.939 Z","level":"error","msg":"can't list merge-request where author in GitLab API","caller":"app/plugin_api.go:940","plugin_id":"com.github.manland.mattermost-plugin-gitlab","err":"GET https://gitlab.com/api/v4/merge_requests: 401 {error: invalid_token}, {error_description: Token is expired. You can either do re-authorization or token refresh.}"}
{"timestamp":"2022-07-08 17:12:47.952 Z","level":"error","msg":"unable to list unreads in GitLab API","caller":"app/plugin_api.go:940","plugin_id":"com.github.manland.mattermost-plugin-gitlab","err":"can't list todo in GitLab api: GET https://gitlab.com/api/v4/todos: 401 {error: invalid_token}, {error_description: Token is expired. You can either do re-authorization or token refresh.}"}
{"timestamp":"2022-07-08 17:12:48.478 Z","level":"debug","msg":"Received HTTP request","caller":"web/handlers.go:156","method":"GET","url":"/api/v4/system/ping","request_id":"uop8nkeei7yyuj5qbgwcfq78or","status_code":"200"}
{"timestamp":"2022-07-08 17:12:48.478 Z","level":"debug","msg":"Received HTTP request","caller":"web/handlers.go:156","method":"GET","url":"/api/v4/system/ping","request_id":"3mracsun4pdpfn9k3ushx5npze","status_code":"200"}
{"timestamp":"2022-07-08 17:12:50.634 Z","level":"debug","msg":"Received HTTP request","caller":"web/handlers.go:156","method":"POST","url":"/api/v4/channels/members/me/view","request_id":"hqedhygnxjngfbzag856q68z6c","status_code":"200"}
{"timestamp":"2022-07-08 17:12:53.478 Z","level":"debug","msg":"Received HTTP request","caller":"web/handlers.go:156","method":"GET","url":"/api/v4/system/ping","request_id":"d84z6eretig19cbzmr6ocq47jc","status_code":"200"}
{"timestamp":"2022-07-08 17:12:54.618 Z","level":"debug","msg":"2022/07/08 17:12:54 [DEBUG] memberlist: Stream connection from=100.127.88.78:35616\n","caller":"cluster/log_parser.go:64","source":"memberlist"}
{"timestamp":"2022-07-08 17:12:54.618 Z","level":"debug","msg":"Received HTTP request","caller":"web/handlers.go:156","method":"GET","url":"/api/v4/analytics/old","request_id":"o8xtqm7npj8xjjsmn9fdnu1xsy","status_code":"200"}
{"timestamp":"2022-07-08 17:12:54.689 Z","level":"debug","msg":"Received HTTP request","caller":"web/handlers.go:156","method":"GET","url":"/api/v4/trial-license/prev","request_id":"zz3354d8wjfbfgzizubzqs5b6r","status_code":"200"}
{"timestamp":"2022-07-08 17:12:55.419 Z","level":"debug","msg":"2022/07/08 17:12:55 [DEBUG] memberlist: Stream connection from=100.127.88.78:35618\n","caller":"cluster/log_parser.go:64","source":"memberlist"}
```
I will post this on https://github.com/mattermost/chimera/issues/10 so we have all the info and discussion in i place.

**Pull request merged**